rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rosbag2_compression/sequential_compression_reader.hpp"
+#include "rosbag2_compression/sequential_compression_writer.hpp"
+#include "rosbag2_cpp/converter_options.hpp"
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/storage_options.hpp"
+#include "rosbag2_cpp/writer.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+#include "rosbag2_storage/ros_helper.hpp"
+#include "rosbag2_storage/storage_filter.hpp"
+#include "rosbag2_storage/topic_metadata.hpp"
+
+namespace rosbag2_py
+{
+
+class Reader
+{
+public:
+  explicit Reader(const std::string & reader_class)
```

Comment:
That sort of logic will hardly scale. I understand that this might result in extra work, which we can address in another PR. But what's needed to support any type of reader implementation?

Reply:
We might be able to add a template argument for the reader class. Looks like pybind11 only supports explicit template instantiations, so we could try something like this: https://github.com/pybind/pybind11/issues/199#issuecomment-323995589

Reply:
In 08db02f, I tried making the subclasses more explicit as opposed to strings. Pybind can export inheritance
https://pybind11.readthedocs.io/en/stable/classes.html#inheritance-and-automatic-downcasting But as in this example and Jacob’s example, the subclasses still need to be explicitly defined as classes for pybind. So this latest version isn’t much better in terms of scaling, and it creates even more work for adding a new reader now, but maybe it is better in terms of making what readers are available explicit. In all three ways of doing it, new readers have to be manually added.

Reply:
See https://github.com/mabelzhang/rosbag2/pull/2/, which uses template specialization instead of defining new classes. It's a tiny bit more compact, and I think makes it a little easier to add new specializations.